### DISCUSSION TITLE
getINITModel2 random error - dispEM

### DISCUSSION THREAD

**Discussion by cherkaos** [SysBioChalmers/Human-GEM#267] [General]
Hi there,

I've been running getINITModel2 and sometimes get this error :
"error using dispEM (line 47)
RAVEN solver not defined or unknown. Try using setRavenSolver('solver')."
It can happen at the beginning or very late after the optimization phase. I always do 'checkInstallation' before and the solver (gurobi) is always detected.

Do you know where this error comes from? 

I am using Gurobi (9.0.0), Matlab (2019b) and it is running on a Linux cluster (CentOS 7.9.2009).

Sarah

**Comment by haowang-bioinfo** [comment]
are you using single-cell data as input?

**Reply by cherkaos** [reply]
No. Normal RNAseq (TPM normalized)

**Comment by JonathanRob** [comment]
Hi @cherkaos, very strange error, I'm not really sure where it may come from. Did you already try running `setRavenSolver('gurobi')` before calling the tINIT function?

**Reply by cherkaos** [reply]
Hi Jonathan

Yes. I actually crashed at the very end, after using the solver (gurobi) multiple times. Attached is the log file.
[log_357.log](https://github.com/SysBioChalmers/Human-GEM/files/6676495/log_357.log)

Also, it worked for most samples but weirdly, all tasks (257 from metabolicTasks_Full.xlsx) are fulfilled. This doesn't make much sense since I didn't set a threshold.

**Reply by JonathanRob** [reply]
Ok, interesting. For the solver issue, maybe because it's being run on a cluster, one of the machines to which the processing is distributed does not have the RAVEN preferences available to it. When you submit your job to the compute cluster, are you processing it on a single node, or across multiple nodes?

As for the tasks, I'm not sure which "threshold" you're referring to. But when you provide tINIT with a task list, it will generate a model that MUST satisfy every one of those tasks (assuming the reference model could perform all the tasks). So if you used the `metabolicTasks_Full.xlsx` file as input for the `getINITModel2` function, then the generated model(s) should be able to fulfill all of the tasks.

Usually, we recommend using the `metabolicTasks_Essential.xlsx` task list as input to tINIT, which is a shorter list of core essential metabolic tasks for cell viability. Only *after* the context-specific model has been generated with tINIT do we then check which tasks from the full list (`metabolicTasks_Full.xlsx`) can be satisfied (i.e., using the `checkTasks` function). This will then result in many of the generated models failing one or more of the tasks from the Full list, since they were not used as constraints in tINIT.

**Reply by cherkaos** [reply]
Ahh no I didn't think of that. I can try to run it again on a single node. 

The threshold is the option to have a hard threshold on expression (the example we used in the tutorial was data_struct.threshold = 1;). I didn't use this option and instead provided the gene expression data from the other tissues instead. That is why I was expecting some tasks to fail.

Okay. If I understand correctly: I will run `getINITModel2`  with the essential tasks. Using the resulting models, I can run  `checkTasks ` to verify if the full tasks are fulfilled. Thank you for the clarification. It is very helpful.

**Reply by JonathanRob** [reply]
Right, if you do not specify a threshold, then tINIT will automatically use the average TPM value across all columns (samples) as the threshold for each gene. For example:

Gene | Sample1 TPM | Sample2 TPM | Sample3 TPM | threshold
----- | -------------- | ------------- | -------------- | ----------
GeneA | 10 | 0 | 2 | 4
GeneB | 1 | 0.5 | 0.75 | 0.75

To clarify the task constraint - if you use the `metabolicTasks_Essential.xslx` task list as input to tINIT, then the resulting tINIT models should all be able to pass all of those tasks. However, they will likely not all pass all the tasks in `metabolicTasks_Full.xlsx`, because the full list includes a lot of tasks not in the smaller essential list. The reason we don't typically use the full task list as input is because many of those tasks are not essential for all cell types to perform, so it doesn't make sense to enforce that as a requirement in the generated models.
